classlib/src/main/java/org/teavm/classlib/java/util/TArrays.java

```diff
@@ -123,92 +131,111 @@ public static <T, U> T[] copyOf(U[] original, int newLength, Class<? extends T[]
         Class<?> componentType = cls.getComponentType();
         T[] result = (T[]) Array.newInstance(componentType, newLength);
         int sz = TMath.min(newLength, original.length);
-        for (int i = 0; i < sz; ++i) {
-            result[i] = (T) componentType.cast(original[i]);
-        }
+        TSystem.arraycopy((TObject) (Object) original, 0, (TObject) (Object) result, 0, sz);
```

Comment:
System?

Reply:
This is made intentionally. In exactly this case no fastArrayCopy optimization should be applied.

Reply:
Why?

Reply:
Take a look at https://github.com/konsoletyper/teavm/pull/868/files#diff-58413b8b823bf4fc7635dfcc153675834256930684714088015dffe99f63ee34R299

Reply:
Ok, if TeaVM optimizes out some check it's not supposed to optimize out, then may be you investigate issue in the optimizer?

Reply:
There is no difference between copyOf(T[] arr, int from, int to) and copyOf(T[] arr, int from, int to, Class<U> cl) from bytecode standpoint. They both are Object[] in bytecode, so optimizer will make no difference. Still, in first case we know that fastArrayCopy is safe while in second case we don't.